### PR TITLE
Use delegate pattern for AutoLocation

### PR DESCRIPTION
**Pull request by tomdale**
**TL;DR** use delegate pattern for autolocation instead of choosing a concrete implementation and returning it from a fake `create()`

Previously, AutoLocation worked by faking out a `create()` method and returning a concrete implementation of either HistoryLocation or AutoLocation.

This was bad for a few reasons. First, messing with `create()` or `extend()` semantics makes it hard for people to reason about behavior. The fact that `AutoLocation.create() instanceof AutoLocation` evaluates to `false` breaks a lot of assumed invariants people have.

Second, it meant that any state needed to pass to the concrete implementation needed to be available at create time, which has forced us to write less-than-elegant code during our app instance refactor for FastBoot.

For example, you cannot say:

``` js
var location = this.container.lookup('location:auto');
location.rootURL = '/foo/';
```

Any state that needs to be set _must_ be set at create time, leading to hard-to-reason-about hacks like setting simple properties via an injection to guarantee it's available at creation time (since `lookup` doesn't take arguments). See https://github.com/emberjs/ember.js/blob/48e115928bcb6b366a621370339354c44aad86b1/packages/ember-routing/lib/system/router.js#L347-L351 for a taste of the madness.

Lastly, concrete implementations are instantiated directly via calls to `create()` rather than going through the container, meaning that `container` does not get set, nor would any registered injections actually get injected.

This commit refactors the entire class, making it a standard Ember.Object, and uses a simple delegate pattern to delegate any calls from the router onto a concrete implementation chosen at runtime.

Because the concrete implementation is not determined until runtime, state can be set on the location at any time until the concrete implementation is accessed, which should allow us to remove some of the contortions we have to do in the current router code.

Note that this commit does not provide passing tests, as all of the current tests are hardcoded to assume the concrete-implementation-at-create-time semantics.

I wanted to submit the PR early to verify that this seems like a reasonable approach to people, and to get many eyes on it as soon as possible as this is a large refactor of a critical part of the system. However, I believe the cleanup here will have cascading cleanup effects throughout the system, and that's always a good thing.

### CODE REVIEW

**Comment by rwjblue** [comment]
@tomdale - Have you confirmed that the refactor works properly on non-history supporting browsers (like IE8 and IE9)?

**Comment by tomdale** [comment]
@rwjblue No, I was hoping someone would do that for me. :D :D :dancer:

File: packages/ember-routing/lib/location/auto_location.js [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
 
-import EmberLocation from "ember-routing/location/api";
-import HistoryLocation from "ember-routing/location/history_location";
-import HashLocation from "ember-routing/location/hash_location";
-import NoneLocation from "ember-routing/location/none_location";
-
+import EmberObject from "ember-runtime/system/object";
 import environment from "ember-metal/environment";
-import { supportsHashChange, supportsHistory } from "ember-routing/location/feature_detect";
+import { supportsHashChange, supportsHistory } from "ember-routing/location/util";
```
Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: Please use multi-line imports here.

```javascript
import {
supportsHashChange,
supportsHistory
} from "ember-routing/location/util";
```

**Comment by tomdale** [comment]
_In theory_ I maintained the existing semantics exactly, but I may certainly have made an error.

File: packages/ember-routing/lib/location/auto_location.js [hunk 2]
```diff
@@ -1,13 +1,12 @@
...
+import { getPath, getOrigin, getHash, getFullPath } from "ember-routing/location/util";
```
Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: Please merge with the prior import:

Please use multi-line imports here.

```javascript
import {
supportsHashChange,
supportsHistory,
getPath,
getOrigin,
getHash,
getFullPath
} from "ember-routing/location/util";
```

File: packages/ember-routing/lib/location/auto_location.js (hunk 1 above)
Comment (tomdale) [inline packages/ember-routing/lib/location/auto_location.js]: Wait, is this really a thing?

Comment (seanpdoyle) [inline packages/ember-routing/lib/location/auto_location.js]: @twokul is this something we can lint for in JSCS?

File: packages/ember-routing/lib/location/auto_location.js [hunk 3]
```diff
@@ -73,285 +86,178 @@ export default {
...
 
-  /**
-    @private
+*/
 
-    Returns the current path as it should appear for HistoryLocation supported
-    browsers. This may very well differ from the real current path (e.g. if it
-    starts off as a hashed URL)
+function detectImplementation(options) {
+  var location = options.location,
```
Comment (stefanpenner) [inline packages/ember-routing/lib/location/auto_location.js]: for assignment statements, 1 var per assignment (this is to be nice to our future debugging experiences)

File: packages/ember-routing/lib/location/api.js [hunk 4]
```diff
@@ -192,15 +193,6 @@ export default {
...
-    // Being nice and not changing
-    var href = (this._location || this.location).href;
-    var hashIndex = href.indexOf('#');
-
-    if (hashIndex === -1) {
-      return '';
-    } else {
-      return href.substr(hashIndex);
-    }
+    return getHash(this._location || this.location);
```
Comment (stefanpenner) [inline packages/ember-routing/lib/location/api.js]: should we deprecate one?

File: packages/ember-routing/lib/location/auto_location.js (hunk 3 above)
Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: This is linted by JSCS also, so there should be failing tests for this..

File: packages/ember-routing/lib/location/auto_location.js [hunk 5]
```diff
@@ -73,285 +86,178 @@ export default {
...
-    var origin = location.origin;
-
-    // Older browsers, especially IE, don't have origin
-    if (!origin) {
-      origin = location.protocol + '//' + location.hostname;
-
-      if (location.port) {
-        origin += ':' + location.port;
-      }
+  concreteImplementation: computed(function() {
```
Comment (stefanpenner) [inline packages/ember-routing/lib/location/auto_location.js]: should this be readOnly?

File: packages/ember-routing/lib/location/auto_location.js (hunk 3 above)
Comment (seanpdoyle) [inline packages/ember-routing/lib/location/auto_location.js]: this was caught by JSCS [here](https://travis-ci.org/emberjs/ember.js/jobs/48524132#L197-L204) but didn't break the build. 

Is that intended or a bug?

Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: @seanpdoyle - Definitely failed a test also: https://travis-ci.org/emberjs/ember.js/jobs/48524132#L592

Comment (seanpdoyle) [inline packages/ember-routing/lib/location/auto_location.js]: Oh ok, awesome. Good catch

**Comment by stefanpenner** [comment]
delegate pattern seems good here.

**Comment by fivetanley** [comment]
@tomdale let me know when it's in a verifiable stage and I can test in IE8/9

File: packages/ember-routing/lib/location/auto_location.js (hunk 1 above)
Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: @seanpdoyle / @twokul - This should absolutely be added to JSCS.

File: packages/ember-routing/lib/location/auto_location.js (hunk 3 above)
Comment (tomdale) [inline packages/ember-routing/lib/location/auto_location.js]: @stefanpenner How does that rule affect destructuring assignment, which is what I would do if I had ES6?

Comment (rwjblue) [inline packages/ember-routing/lib/location/auto_location.js]: The rule is specifically to aid in debugging.  Consider the following example:

```javascript
var foo = this.blah(),
    bar = this._someThing,
    baz = someOtherFunction(foo);
```

If `someOtherFunction` is throwing an error (or maybe you need to check the output of `foo` before calling `someOtherFunction`), how do you pause at that section without manually stepping into/over functions that are not related?

@tomdale - I would **absolutely love** more involvement in choosing the style guide directives that we go with (and automate).  Please submit a PR to the style guide and `.jscsrc` so we can discuss the merits of specific changes.

Comment (stefanpenner) [inline packages/ember-routing/lib/location/auto_location.js]: > @stefanpenner How does that rule affect destructuring assignment, which is what I would do if I had ES6?
simply expands obvious structures, like above. That is why i said, this scenario like doesn't matter, just a good opportunity for to trick you into giving feedback on the style guide ;)

```js
let [a,b,c] = array;
let ({d,e,f}) = obj;
````

were it does matter is

```js
var a = doSomething();
var b = doSomethingElse();
var c = doSomethingVeryDifferent();
```

Comment (tomdale) [inline packages/ember-routing/lib/location/auto_location.js]: > trick you into giving feedback on the style guide ;)

Literally LOL. Agree with you on the debugging stuff, sounds good. :+1:

File: packages/ember-routing/lib/location/api.js (hunk 4 above)
Comment (tomdale) [inline packages/ember-routing/lib/location/api.js]: I don't think we need to deprecate, it's private and I can just update. Not sure why that wasn't done before.

Comment (stefanpenner) [inline packages/ember-routing/lib/location/api.js]: if its private, lets just drop whichever you prefer least.

**Comment by stefanpenner** [comment]
@fivetanley thanks!

File: packages/ember-routing/lib/location/auto_location.js (hunk 1 above)
Comment (twokul) [inline packages/ember-routing/lib/location/auto_location.js]: @seanpdoyle @rwjblue absolutely

File: packages/ember-routing/lib/location/api.js (hunk 4 above)
Comment (tomdale) [inline packages/ember-routing/lib/location/api.js]: Done.

**Comment by tomdale** [comment]
@fivetanley This should be good to test now.

fark_nrdv-i6zvuadw8lvcqbb0zur-gk

**Comment by SamSaffron** [comment]
will the new API mean we don't need this kind of hoop jumping? https://github.com/discourse/discourse/blob/stable/app/assets/javascripts/discourse/routes/discourse_location.js 

cc @eviltrout

**Comment by eviltrout** [comment]
@SamSaffron I suspect we could just overwrite the bits we need yes, so we could save a bunch of LOC. Looks good @tomdale

**Comment by tomdale** [comment]
@SamSaffron @eviltrout One thing to keep in mind here is that this refactor makes AutoLocation's lookup of HashLocation and HistoryLocation go through the container, and thus the resolver. That means, if you do need to override any behavior, you can just subclass the respective concrete implementation and put it in the right place in the filesystem.

So for example, if you had the location set to `'auto'` but wanted to tweak how hash routing worked, just do:

```js
// app/locations/hash.js
export default Ember.HashLocation.extend({
  init: function() {
    this._super.apply(this, arguments);
    this.doCrazyStuff();
  }
  // ...
});
```

This subclass should now get picked up by the auto location automatically.

**Comment by fivetanley** [comment]
@tomdale I can take a look at this first thing tomorrow morning. Need to redownload the ie9 vm.

**Comment by tomdale** [comment]
@fivetanley!!!!!!!!!

**Comment by fivetanley** [comment]
i was gonna test in ie9 but then i got high

seriously tonight right after work though

**Comment by tomdale** [comment]
@fivetanley

**Comment by tomdale** [comment]
op-will-surely-deliver-lets-just-wait jpg original

**Comment by fivetanley** [comment]
@tomdale I couldn't build this locally due to https://github.com/emberjs/ember.js/pull/10380/. Can you rebase/merge master? I wasn't sure which changes to keep when resolving the conflicts. The conflicts mostly seem to be renamed or added tests.

**Comment by fivetanley** [comment]
I am very confused as to why the rebase shows me as the committer for @rwjblue's and @ef4's commits. Anyone have any advice?

Otherwise, this passes in IE8 and IE9 (aside to an unrelated indexOf call)

**Comment by ef4** [comment]
That would happen if you had already merged our commits before rebasing.

**Comment by fivetanley** [comment]
Ah, that was it. Thanks @ef4! Reset and repushed, I expect this to pass CI.
